aspnetcore/blazor/components/prerender.md

```diff
@@ -17,358 +17,134 @@ uid: blazor/components/prerender
-
-    void IDisposable.Dispose() => persistingSubscription.Dispose();
 }
 ```
 
-When the component executes, `currentCount` is only set once during prerendering. The value is restored when the component is rerendered. The following is example output.
+However, the preceding approach isn't useful if your logic requires a value during prerendering.
 
-> [!NOTE]
-> If the app adopts [interactive routing](xref:blazor/fundamentals/routing#static-versus-interactive-routing) and the page is reached via an internal [enhanced navigation](xref:blazor/fundamentals/routing#enhanced-navigation-and-form-handling), prerendering doesn't occur. Therefore, you must perform a full page reload for the component to see the following output. For more information, see the [Interactive routing and prerendering](#interactive-routing-and-prerendering) section.
-
-> :::no-loc text="info: BlazorSample.Components.Pages.PrerenderedCounter3[0]":::  
-> :::no-loc text="      currentCount set to 96":::  
-> :::no-loc text="info: BlazorSample.Components.Pages.PrerenderedCounter3[0]":::  
-> :::no-loc text="      currentCount restored to 96":::
-
-:::moniker-end
+You can also avoid the problem if you [disable prerendering](#disable-prerendering) for the component, but that's an extreme measure to take in many cases that may not meet your component's specifications.
 
-:::moniker range="< aspnetcore-10.0"
+There are a three approaches that you can take to address this scenario. The following are listed from most recommended to least recommended:
 
-To preserve prerendered state, decide what state to persist using the <xref:Microsoft.AspNetCore.Components.PersistentComponentState> service. <xref:Microsoft.AspNetCore.Components.PersistentComponentState.RegisterOnPersisting%2A?displayProperty=nameWithType> registers a callback to persist the component state during prerendering. The state is retrieved when the component renders interactively. Make the call at the end of initialization code in order to avoid a potential race condition during app shutdown.
+* *Recommended* for shared framework services: For shared framework services that merely aren't registered server-side in the main project, register the services in the main project, which makes them available during prerendering. For an example of this scenario, see the guidance for <xref:System.Net.Http.HttpClient> services in <xref:blazor/call-web-api?pivots=webassembly#client-side-services-for-httpclient-fail-during-prerendering>.
 
-The following counter component example persists counter state during prerendering and retrieves the state to initialize the component.
+* *Recommended* for services outside of the shared framework: Create a custom service implementation for the service on the server. Use the service normally in interactive components of the `.Client` project. For a demonstration of this approach, see <xref:blazor/fundamentals/environments#read-the-environment-client-side-in-a-blazor-web-app>.
```

Comment:
Implementing `IWebAssemblyHostEnvironment` on the server seems kinda icky to me, but it does work. I didn't realize we had so much content about how to get the environment from a component in a hosting model agnostic way. Maybe we should just address this in the framework.